bench/macro/rocksdb-bench-wp8.hs

```diff
@@ -0,0 +1,418 @@
+
+-------------------------------------------------------------------------------
+-- command line interface
+-------------------------------------------------------------------------------
+
+globalOptsP :: O.Parser GlobalOpts
+globalOptsP = pure GlobalOpts
+    <*> pure "_bench_rocksdb"
+    <*> O.option O.auto (O.long "initial-size" <> O.value 100_000_000 <> O.showDefault <> O.help "Initial LSM tree size")
+
+cmdP :: O.Parser Cmd
+cmdP = O.subparser $ mconcat
+    [ O.command "setup" $ O.info
+        (CmdSetup <$> setupOptsP <**> O.helper)
+        (O.progDesc "Setup benchmark")
+
+    , O.command "dry-run" $ O.info
+        (CmdDryRun <$> runOptsP <**> O.helper)
+        (O.progDesc "Dry run, measure overhead")
+
+    , O.command "run" $ O.info
+        (CmdRun <$> runOptsP <**> O.helper)
+        (O.progDesc "Proper run")
+    ]
+
+setupOptsP :: O.Parser SetupOpts
+setupOptsP = pure SetupOpts
+
+runOptsP :: O.Parser RunOpts
+runOptsP = pure RunOpts
+    <*> O.option O.auto (O.long "batch-count" <> O.value 1000 <> O.showDefault <> O.help "Batch count")
+    <*> O.option O.auto (O.long "batch-size" <> O.value 256 <> O.showDefault <> O.help "Batch size")
+    <*> O.switch (O.long "check" <> O.help "Check generated key distribution")
+    <*> O.option O.auto (O.long "seed" <> O.value 1337 <> O.showDefault <> O.help "Random seed")
+
+-------------------------------------------------------------------------------
+-- clock
+-------------------------------------------------------------------------------
+
+timed :: IO a -> IO (a, Double)
+timed action = do
+    t1 <- Clock.getTime Clock.Monotonic
+    x  <- action
+    t2 <- Clock.getTime Clock.Monotonic
+    let !t = fromIntegral (Clock.toNanoSecs (Clock.diffTimeSpec t2 t1)) * 1e-9
+    return (x, t)
+
+timed_ :: IO () -> IO Double
+timed_ action = do
+    t1 <- Clock.getTime Clock.Monotonic
+    action
+    t2 <- Clock.getTime Clock.Monotonic
+    return $! fromIntegral (Clock.toNanoSecs (Clock.diffTimeSpec t2 t1)) * 1e-9
+
+-------------------------------------------------------------------------------
+-- setup
+-------------------------------------------------------------------------------
+
+{-
+rocksdb::BlockBasedTableOptions table_options;
+  table_options.filter_policy.reset(rocksdb::NewBloomFilterPolicy(10, false));
+  my_cf_options.table_factory.reset(
+      rocksdb::NewBlockBasedTableFactory(table_options));
+-}
+
+rocksDbWithOptions :: (RocksDB.Options -> IO r) -> IO r
+rocksDbWithOptions kont =
+    RocksDB.withOptions $ \options ->
+    RocksDB.withFilterPolicyBloom 10 $ \filterPolicy ->
+    RocksDB.withBlockTableOptions $ \topts -> do
+        RocksDB.blockBasedOptionsSetFilterPolicy topts filterPolicy
+
+        RocksDB.optionsSetCreateIfMissing options True
+        RocksDB.optionsIncreaseParallelism options 4 -- the benchmark spec says to use one core though.
+        RocksDB.optionsOptimizeLevelStyleCompaction options 0x200_00000 -- 32*1024*1024; default is 512*1024*1024?
```

Comment:
What does this option do?

Reply:
https://github.com/facebook/rocksdb/wiki/Compaction#overview-of-compaction-algorithms